src/components/FileInput.js

```diff
@@ -1,34 +1,34 @@
-import React, { PureComponent } from 'react'
-import propTypes from 'prop-types'
+import React, { Component } from 'react'
+import propTypes from '@dhis2/prop-types'
 import i18n from '@dhis2/d2-i18n'
-import { FileInput as UiCoreFileInput, Help, Field } from '@dhis2/ui-core'
+import {
+    FileInputField as UiCoreFileInputField,
+    FileListItem,
+} from '@dhis2/ui-core'
 import { FieldAdapter, adapterComponentProps } from './FieldAdapter.js'
 
-const messages = {
-    uploadFile: i18n.t('Upload file'),
-    uploadFiles: i18n.t('Upload files'),
-}
+const btnLabel = i18n.t('Upload file')
+const btnLabelMulti = i18n.t('Upload files')
 
-class FileInput extends PureComponent {
-    onFileInputChange = fileList => {
+class FileInputComponent extends Component {
```

Comment:
Does this component have to be a class based component? If possible I'd personally prefer a functional component here.

Reply:
Yeah me too, I mentioned it under my first "point of attention" in the PR description:
> FileInput is still a class component because I'd need an onRemove callback on the FileListItems which are being mapped over. I believe hooks have some trouble in these type of scenarios. But I'd prefer to hook this component up too if possible.

I suppose I could just create a new function on every iteration.... Any advise is very welcome.

Reply:
```jsx
const onFileInputChange = ({ value, multifile, onChange }) = ({ files }) => {
  const newValue = multifile ? dedupeAndConcat(files) : [...files]
  onChange({ value: newValue })
}

const onFileRemove = ({ value, onChange }) => fileToDelete => {
  const files = value.filter(file => file !== fileToDelete)
  onChange(files.length > 0 ? files : '')
}

const dedupeAndConcat = (oldValue, fileList) => {
  return [...value, ...fileList].reduceRight(
    (acc, file) => {
      if (!acc.unique.has(file.name)) {
        acc.unique.add(file.name)
        acc.files.unshift(file)
      }
      return acc
    },
    { files: [], unique: new Set() }
  ).files
}

const FileInputComponent = props => {
  const buttonLabel = props.buttonLabel || props.multifile ? props.btnLabelMulti : props.btnLabel
  const disabled = = disabled || (!multifile && value.length >= 1)

  return (
    <UiCoreFileInputField
      {...props}
      onChange={onFileInputChange(props)}
      buttonLabel={buttonLabel}
      disabled={disabled}
    >
      {Array.isArray(value)
        ? value.map(file => (
          <FileListItem
            key={file.name}
            label={file.name}
            onRemove={onFileRemove(props)}
            removeText={i18n.t('Remove')}
          />
          ))
        : undefined}
    </UiCoreFileInputField>
  )
}
```

This should be it, more or less, right? (untested)